### PR TITLE
Write `Package.from_ipfs` class method

### PR DESCRIPTION
**Pull request by njgheorghita**
### What was wrong?
Lacked functionality to instantiate a `Package` from an IPFS uri

### How was it fixed?
Jacked some utils from Populus, and added a class method (`from_ipfs`) to `Package` class that creates a `Package` if the IPFS uri resolves to a valid Package

#### Cute Animal Picture
image

### CODE REVIEW

File: ethpm/utils/ipfs.py [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+
+
+INFURA_IPFS_URI_PREFIX = "https://ipfs.infura.io/ipfs/"
+
+
+def fetch_ipfs_package(hash: str) -> Dict[str, Any]:
+    package_uri = INFURA_IPFS_URI_PREFIX + hash
+    with request.urlopen(package_uri) as response:
+        package_data = response.read()
+    return json.loads(to_text(package_data))
```
Comment (pipermerriam) [inline ethpm/utils/ipfs.py]: I think you can get this as `response.json()`

You also *probably* should:

1. call `response.raise_for_status()` to take care of error http responses.
2. not MVP but worth tracking as an issue and maybe commenting in the code as well, this should eventually be turned into a swappable *backend* system so that people can use something other than the infura IPFS gateway.

File: tests/ethpm/test_package_ipfs_initialization.py [hunk 2]
```diff
@@ -0,0 +1,27 @@
+import pytest
+
+from ethpm import Package
+
+
+VALID_IPFS_PKG = 'ipfs://QmeD2s7KaBUoGYTP1eutHBmBkMMMoycdfiyGMx2DKrWXyV'
+
+
+def test_package_from_ipfs_with_valid_uri():
+    package = Package.from_ipfs(VALID_IPFS_PKG)
```
Comment (pipermerriam) [inline tests/ethpm/test_package_ipfs_initialization.py]: It appears this is actually making the http request.  This is generally considered an anti-pattern.  Instead, I'd recommend using the `monkeypatch` pytest fixture to mock out the `fetch_ipfs_package` function with a canned response.

File: tests/ethpm/test_package_ipfs_initialization.py [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+        return valid_package
+    monkeypatch.setattr(ipfs, 'fetch_ipfs_package', mock_fetch)
+
+
+def test_package_from_ipfs_with_valid_uri(monkeypatch, valid_package):
+    xxx = ipfs.fetch_ipfs_package(VALID_IPFS_PKG)
+    package = Package.from_ipfs(VALID_IPFS_PKG)
+    print(xxx)
+    print(package)
+    assert isinstance(package, str)
```
[the remark this answers is not in the document]
Comment (pipermerriam) [reply tests/ethpm/test_package_ipfs_initialization.py]: Try monkeypatching `ethpm.package.fetch_ipfs_package`.

[the remark this answers is not in the document]
Comment (njgheorghita) [reply tests/ethpm/test_package_ipfs_initialization.py]: Can you explain a little more about why that's possible? 
Every time you call `fetch_ipfs_package` in the `package` module, you'll use the imported version (available at the module level), which is a copy of and not a reference to `ipfs.fetch_ipfs_package` ??

[the remark this answers is not in the document]
Comment (pipermerriam) [reply tests/ethpm/test_package_ipfs_initialization.py]: So `ethpm.package.fetch_ipfs_package` is a reference to `ethpm.utils.ipfs.fetch_ipfs_package`.  When you monkeypatch the later, everything that *already* had a reference to it retains that reference.  It's basically a race.  *if* you were able to do the monkeypatch *before* loading the `epthm.package` module, then `ethpm.package.fetch_ipfs_package` would end up being the monkeypatched version (and would remain the monkeypatched version *after* the monkeypatch had been removed.

You can get around this by not importing it directly, but rather importing the module.

```python
# ethpm/package.py
from ethpm.utils import ipfs

def thing(...):
    ipfs.fetch_ipfs_package(...)
```

In this scenario, the `package` module resolves `ipfs.fetch_ipfs_package` everytime it uses it, and thus, if you monkeypatch it, then the `package` module will use the monkeypatched version.

Basically, python is mutable as hell and it gets really confusing when we start mutating modules like this.

[the remark this answers is not in the document]
Comment (njgheorghita) [reply tests/ethpm/test_package_ipfs_initialization.py]: Cool, makes sense. thanks for the explanation!
